### PR TITLE
Application Timeline - My Fit step (Long-form questions)

### LINKED ISSUE #2888: Task - Application - UI for Long-form Questions step of the Application
# Description
Create UI in react for filling out any number of long-form application questions. Each answer textarea should include a word-counter component.

Though the wireframe includes a Skill Questions section with an associated skill, we can ignore this for now.

# Images
https://xd.adobe.com/view/a678f763-f7b7-43ca-740a-89359cbe42f1-d049/screen/3bd349c0-6dff-49ae-abad-70b1d27c0f43/Step-2B-Requirements-7

### PR DESCRIPTION
**Pull request by yonikid15**
Resolves #2888 

### Notes:
- To accomplish saving all questions individually as well as having a "Submit All" button I used reacts reference hooks. Formik comes with a `innerRef` prop for that purpose. This allows triggering submission (as well as access to other Formik props) outside the form. 
- Currently, the "Save All" button triggers the submission on all the individual forms. It's probably better to create an array of all the answers and make a single request instead. This shouldn't be hard to implement.

### CODE REVIEW

File: resources/assets/js/models/types.ts [hunk 1]
```diff
@@ -172,12 +172,26 @@ export interface Job {
...
   id: number;
   job_poster_id: number;
   description: localizedFieldNonNull;
 }
 
+export interface JobPosterQuestion {
+  id: number;
+  job_poster_id: number;
+  question: localizedFieldNonNull;
+  description: localizedFieldNonNull;
```
Comment (tristan-orourke) [inline resources/assets/js/models/types.ts:192]: These fields ARE nullable, so use `localizedField` instead. Generally, most content entered by a user (including a manager) can be null. `localizedFieldNonNull` should mostly be used for constants that are only created by devs or admins, like Skills or Provinces.

File: resources/assets/js/components/Application/MyFit/MyFit.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+          id="application.myfit.firstParagraph"
+          defaultMessage="The manager has included a few further questions that they'd like you to answer in order to be considered for this position. Questions that require an answer are marked, while all others are optional."
+          description="Paragraph for the My Fit step in the Application Timeline"
+        />
+      </p>
+      {jobQuestions.map((question, index) => {
+        formRefs.current = [
+          ...formRefs.current,
+          formRefs.current[index] || createRef(),
+        ];
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply resources/assets/js/components/Application/MyFit/MyFit.tsx:123]: Yup! I saw that during review yesterday and made a fix.

File: resources/assets/js/components/Application/MyFit/MyFit.tsx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        ref.current.submitForm(),
+      ),
+    );
+  };
+
+  const validateAndSubmit = (
+    refs: React.MutableRefObject<
+      React.MutableRefObject<FormikProps<QuestionValues>>[]
+    >,
+    nextStep: string,
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply resources/assets/js/components/Application/MyFit/MyFit.tsx:88]: This is supposed to be `nextStepUrl` whenever we establish what the routes will be.

File: resources/assets/js/components/Application/MyFit/MyFit.tsx [hunk 4]
```diff
@@ -117,10 +116,12 @@ export const MyFit: React.FunctionComponent<MyFitProps> = ({
...
-        formRefs.current = [
-          ...formRefs.current,
-          formRefs.current[index] || createRef(),
-        ];
+        if (!formRefs.current.includes(formRefs.current[index])) {
+          formRefs.current = [
+            ...formRefs.current,
+            formRefs.current[index] || createRef(),
+          ];
+        }
```
Comment (tristan-orourke) [inline resources/assets/js/components/Application/MyFit/MyFit.tsx:69]: This doesn't seem quite right. The if statement asks is `current` includes an element of `current` - looks like circular logic. And I'm not sure its safe to use index as the key. If a question is deleted, the same index may later refer to a different question. Maybe using question.id would be better?

File: resources/assets/js/helpers/forms.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import { FormikProps, FormikValues } from "formik";
+import _ from "lodash";
```
Comment (tristan-orourke) [inline resources/assets/js/helpers/forms.ts:2]: You can import only the function you need for a smaller file, like this: `import isEmpty from "lodash/isEmpty";`

File: resources/assets/js/models/types.ts [hunk 6]
```diff
@@ -172,12 +172,26 @@ export interface Job {
   culture_special: localizedField;
 }
 
+export interface JobApplicationAnswer {
+  id: number;
+  job_poster_questions_id: number;
+  job_application_id: number;
+  answer: localizedField;
```
Comment (tristan-orourke) [inline resources/assets/js/models/types.ts:179]: Should be just a string.

File: resources/assets/js/components/Application/MyFit/Question.tsx [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+interface QuestionProps {
+  jobApplicationAnswer: JobApplicationAnswer;
+  formRef: any;
+  index: number;
+  question: JobPosterQuestion;
+  handleSubmit: (data: JobApplicationAnswer) => Promise<void>;
+}
+
+export interface QuestionValues {
+  id: number;
```
Comment (tristan-orourke) [inline resources/assets/js/components/Application/MyFit/Question.tsx:34]: If this is just the Form Values, I think the only property necessary here is the `answer` field.

File: resources/assets/js/helpers/forms.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+export const focusOnElement = (id: string): void => {
+  const element = document.getElementById(id);
+  if (element) {
+    element.focus();
+  }
+};
+
+export const validateAllForms = async (
+  refs: MutableRefObject<MutableRefObject<FormikProps<FormikValues>>[]>,
```
Comment (tristan-orourke) [inline resources/assets/js/helpers/forms.ts:13]: I realized while poking around that in this function and submitAllForms, we just need the inner array of refs, there's no need to pass in the array wrapped in another MutableRefObject.

Comment (yonikid15) [reply resources/assets/js/helpers/forms.ts]: Noted. Thanks again for the help. I'm going to keep looking for a solution to the `focusOnElement` issue tomorrow.

File: resources/assets/js/components/Application/MyFit/MyFit.tsx [hunk 9]
```diff
@@ -116,11 +115,8 @@ export const MyFit: React.FunctionComponent<MyFitProps> = ({
               data-c-radius="rounded"
               type="button"
               disabled={isSubmitting}
-              onClick={() =>
-                validateAndSubmit(
-                  formRefs.current,
-                  "ReplaceWithUrlOfPreviousStep",
-                )
+              onClick={(): Promise<void> =>
+                validateAndSubmit(formRefs.current).then(handleReturn)
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply resources/assets/js/components/Application/MyFit/MyFit.tsx:119]: This was actually your suggestion on friday :p I just forgot to push it up.

[the remark this answers is not in the document]
Comment (tristan-orourke) [reply resources/assets/js/components/Application/MyFit/MyFit.tsx:123]: Aw, after testing it this doesn't quite work yet. First of all, the handle______ functions trigger even if validation or submission fails. But also, the handle____ functions are triggering _before_ the handleSubmit callback completes.

**Pull request merged**